File: advent-of-code/day03/javascript/two.js
const fs = require("fs");
const path = require("path");

const inputPath = path.join(__dirname, "..", "input.txt");
const input = fs.readFileSync(inputPath, "utf-8").trim().split("\n");

const findAndAddLargestTwelveDigits = (input) => {
  const largestTwelveDigits = [];

  for (const line of input) {
    const digits = line.split("").map(Number);
    const largestDigits = Array.from({ length: 12 }, () => 0);
    const indexOfLargestDigits = Array.from({ length: 12 }, () => 0);

    for (let i = 0; i <= 11; i++) {
      const indexOfPreviousLargestDigit =
        i === 0 ? -1 : indexOfLargestDigits[i - 1];
      const start = indexOfPreviousLargestDigit + 1;
      const end = digits.length - (11 - i);
      for (let j = start; j < end; j++) {
        if (digits[j] > largestDigits[i]) {
          largestDigits[i] = digits[j];
          indexOfLargestDigits[i] = j;
        }
      }
    }
    largestTwelveDigits.push(Number(largestDigits.join("")));
  }

  console.log("Largest twelve digits from each line:", largestTwelveDigits);
  const sumOfLargestTwelveDigits = largestTwelveDigits.reduce(
    (acc, curr) => acc + curr,
    0
  );
  console.log("Sum of largest twelve digits:", sumOfLargestTwelveDigits);
  return largestTwelveDigits;
};

findAndAddLargestTwelveDigits(input);

File: advent-of-code/day03/python/two.py
import os

input_path = os.path.join(os.path.dirname(__file__), "..", "input.txt")

def find_and_add_twelve_largest_digits() -> int:
    largest_twelve_digits = []

    with open(input_path, "r") as file:
        input = file.readlines()

        for line in input:
            digits = [int(char) for char in line.strip()]
            largest_digits = [0] * 12
            index_of_largest_digits = [0] * 12

            for i in range(12):
                index_of_previous_largest = -1 if i == 0 else index_of_largest_digits[i - 1]
                start = index_of_previous_largest + 1
                end = len(digits) - (11 - i)
                for j in range(start, end):
                    if digits[j] > largest_digits[i]:
                        largest_digits[i] = digits[j]
                        index_of_largest_digits[i] = j

            print("Largest twelve digits in line:", largest_digits)
            largest_twelve_digits.extend([int(''.join(map(str, largest_digits)))])

    print("All largest twelve digits from all lines:", largest_twelve_digits)
    sum_largest_twelve_digits = sum(largest_twelve_digits)

    print("Sum of all largest twelve digits:", sum_largest_twelve_digits)
    return sum_largest_twelve_digits

find_and_add_twelve_largest_digits()